### PR TITLE
finalize tournament lobby +  add variable event cards

### LINKED ISSUE #909: verify isLobbyOpen logic
currently seems to be reversing before/after offsets

### PR DESCRIPTION
**Pull request by sgfost**
- [x] resolves virtualcommons/planning#51
- [x] resolves virtualcommons/planning#44
- [x] resolves virtualcommons/planning#52
- [x] resolves #909

### CODE REVIEW

File: server/src/rooms/lobby/tournament/index.ts [hunk 1]
```diff
@@ -27,19 +33,56 @@ export class TournamentLobbyRoom extends LobbyRoom<TournamentLobbyRoomState> {
...
+    if (timeout) {
+      clearTimeout(timeout);
+      this.groupTimeouts.delete(groupId);
+    }
+  }
+
+  formGroup(): string {
+    const group = [];
+    const groupId = uuidv4();
+    _.shuffle(this.state.queue);
```
Comment (alee) [inline server/src/rooms/lobby/tournament/index.ts:78]: would it be more straightforward to shuffle once and drain all valid groups at once and then leave the unlucky leftovers in the queue?

File: server/src/rooms/lobby/tournament/state.ts [hunk 2]
```diff
@@ -12,49 +9,51 @@ const logger = settings.logging.getLogger(__filename);
 export class TournamentLobbyRoomState extends LobbyRoomState {
   maxClients = 200;
 
-  clientsToGroups = new Map<string, LobbyClient[]>();
+  // only holds clients that still need to be put into a group
+  queue: LobbyClient[] = [];
+  // groups indexed by a unique group id
+  pendingGroups: Map<string, LobbyClient[]> = new Map();
+  // map client username to group id
+  clientToGroup: Map<string, string> = new Map();
```
Comment (alee) [inline server/src/rooms/lobby/tournament/state.ts:17]: Would a `Group` class with a fixed `size`, `LobbyClient[]` and `id` simplify things? Then `clientToGroup = Map<string, Group>`

Comment (sgfost) [reply server/src/rooms/lobby/tournament/state.ts:17]: Almost certainly, could also just store the group reference on the LobbyClient which I think was how the old tournament lobby did it but that'll either be a case of a really messy extension of the LobbyClient schema class or putting it in the base class to be a dead property elsewhere. The map is probably fine.. only needed for one operation

File: server/src/rooms/lobby/tournament/index.ts (hunk 1 above)
Comment (sgfost) [reply server/src/rooms/lobby/tournament/index.ts:78]: Yeah, the queue isn't actually managed correctly, I'll look at this when doing so

**Comment by sgfost** [comment]
Unit tests for the grouping logic + a fair bit of mostly manual smoke-testing have me feeling fairly confident about the lobby working correctly. I'll move onto the other few requirements in order to finish up this week and leave the start of next for further testing

**Comment by sgfost** [comment]
the problem with treatments during the pretest was that I somehow botched applying the changes in [14f8bfd](https://github.com/virtualcommons/port-of-mars/pull/908/commits/14f8bfddb3a7d93c432e08f786da6b7c21d60cb0) which fixes the issue. Fix for the lobby not staying open long enough is correct as well. Better tests added for both

**Review by alee — APPROVED** [review]
great work as always @sgfost !

**Pull request merged**